Source/DafnyStandardLibraries/src/Std/Actions/Actions.dfy

```diff
@@ -0,0 +1,620 @@
+      ensures Action().CanConsume(history, next)
+  }
+
+  ghost predicate OnlyProduces<T, R>(i: Action<T, R>, history: seq<(T, R)>, c: R) {
+    i.CanProduce(history) <==> forall e <- history :: e.1 == c
+  }
+
+  ghost predicate Terminated<T>(s: seq<T>, stop: T -> bool, n: nat) {
+    forall i | 0 <= i < |s| :: n <= i <==> stop(s[i])
+  }
+
+  lemma TerminatedDistributes<T>(left: seq<T>, right: seq<T>, stop: T -> bool, n: nat)
+    requires Terminated(left, stop, |left|)
+    requires Terminated(right, stop, n)
+    ensures Terminated(left + right, stop, |left| + n)
+  {}
+
+  lemma TerminatedUndistributes<T>(left: seq<T>, right: seq<T>, stop: T -> bool, n: nat)
+    requires Terminated(left + right, stop, n)
+    ensures Terminated(left, stop, n)
+    ensures Terminated(right, stop, Max(0, n - |left|))
+  {
+    assert forall i | 0 <= i < |left| :: left[i] == (left + right)[i];
+    assert forall i | 0 <= i < |right| :: right[i] == (left + right)[i + |left|];
+  }
+
+  trait ProducesTerminatedProof<T, R> extends ConsumesAllProof<T, R> {
+
+    ghost function FixedInput(): T
+    ghost function StopFn(): R -> bool
+    ghost function Limit(): nat
+
+    lemma ProducesTerminated(history: seq<(T, R)>)
+      requires Action().CanProduce(history)
+      requires (forall i <- Inputs(history) :: i == FixedInput())
+      ensures exists n: nat | n <= Limit() :: Terminated(Outputs(history), StopFn(), n)
+
+    // Termination metric
+    ghost function Remaining(): nat
+      requires Action().Valid()
+      requires forall i <- Action().Consumed() :: i == FixedInput()
+      reads Action().Repr
+    {
+      ProducesTerminated(Action().history);
+      var n: nat :| n <= Limit() && Terminated(Action().Produced(), StopFn(), n);
+      TerminatedDefinesNonTerminalCount(Action().Produced(), StopFn(), n);
+      Limit() - NonTerminalCount(Action().Produced(), StopFn())
+    }
+
+    twostate lemma InvokeUntilTerminationMetricDecreased(new nextProduced: R)
+      requires old(Action().Valid())
+      requires Action().Valid()
+      requires forall i <- old(Action().Consumed()) :: i == FixedInput()
+      requires Action().Consumed() == old(Action().Consumed()) + [FixedInput()]
+      requires Action().Produced() == old(Action().Produced()) + [nextProduced]
+      requires !StopFn()(nextProduced)
+      ensures Remaining() < old(Remaining())
+    {
+      var before := old(Action().Produced());
+      var after := Action().Produced();
+      ProducesTerminated(old(Action().history));
+      var n: nat :| n <= Limit() && Terminated(before, StopFn(), n);
+      ProducesTerminated(Action().history);
+      var m: nat :| m <= Limit() && Terminated(after, StopFn(), m);
+      if n < |before| {
+        assert StopFn()(before[|before| - 1]);
+        assert !StopFn()(Action().Produced()[|Action().Produced()| - 1]);
+        assert |Action().Produced()| <= m;
+        assert !StopFn()(Action().Produced()[|before| - 1]);
+        assert false;
```

Comment:
Alternatively, this could be formulated as

``` dafny
assert false by {
  // the other 4 asserts
}
```

Reply:
I definitely like that better, puts the conclusion right up front. Otherwise it's more like you're waiting to the end of the movie to find out "Aha, he's actually been a ghost this whole time!" :)